cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,1058 @@
+    auto nr_uniqe_edge_types_partition =
+      thrust::count_if(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator(size_t{0}),
+                       thrust::make_counting_iterator(edgelist_types.size()),
+                       detail::is_first_in_run_t<edge_type_t const*>{edgelist_types.data()});
+
+    rmm::device_uvector<edge_type_t> unique_types(nr_uniqe_edge_types_partition,
+                                                  handle.get_stream());
+    rmm::device_uvector<edge_t> type_offsets(nr_uniqe_edge_types_partition + 1,
+                                             handle.get_stream());
+
+    thrust::fill(handle.get_thrust_policy(), type_offsets.begin(), type_offsets.end(), edge_t{0});
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          edgelist_types.begin(),
+                          edgelist_types.end(),
+                          thrust::make_constant_iterator(size_t{1}),
+                          unique_types.begin(),
+                          type_offsets.begin());
+
+    thrust::exclusive_scan(handle.get_thrust_policy(),
+                           type_offsets.begin(),
+                           type_offsets.end(),
+                           type_offsets.begin(),
+                           size_t{0});
+
+    std::vector<edge_type_t> h_unique_types(unique_types.size());
+    std::vector<edge_t> h_type_offsets(type_offsets.size());
+
+    raft::update_host(
+      h_unique_types.data(), unique_types.data(), unique_types.size(), handle.get_stream());
+
+    raft::update_host(
+      h_type_offsets.data(), type_offsets.data(), type_offsets.size(), handle.get_stream());
+
+    for (size_t idx = 0; idx < h_unique_types.size(); idx++) {
+      auto typ                   = h_unique_types[idx];
+      auto nr_elements_to_insert = (h_type_offsets[idx + 1] - h_type_offsets[idx]);
+
+      auto values_to_insert =
+        cugraph::allocate_dataframe_buffer<value_t>(nr_elements_to_insert, handle.get_stream());
+
+      auto zip_itr = thrust::make_zip_iterator(
+        thrust::make_tuple(edgelist_majors.begin(), edgelist_minors.begin()));
+
+      // FIXME: Can we avoid copying
+      thrust::copy(handle.get_thrust_policy(),
+                   zip_itr + h_type_offsets[idx],
+                   zip_itr + h_type_offsets[idx] + nr_elements_to_insert,
+                   cugraph::get_dataframe_buffer_begin(values_to_insert));
+
+      static_assert(std::is_same_v<
+                    typename thrust::iterator_traits<decltype(cugraph::get_dataframe_buffer_begin(
+                      values_to_insert))>::value_type,
+                    value_t>);
+
+      search_container.insert(handle,
+                              typ,
+                              raft::device_span<edge_t>(edgelist_ids.begin() + h_type_offsets[idx],
+                                                        nr_elements_to_insert),
+                              std::move(values_to_insert));
+    }
```

Comment:
Let me spend more time on deciphering this code but my gut impression is that, this code is more complex than it needs to be.

Reply:
Shouldn't we just decompress to (edge_id, type, src, dst) tuples, shuffle, and insert?